build.Containerfile: reproducible reference build via buildah/podman/docker

schildbach opened the pull request:
- FIXME the JavaDoc doesn't seem to be reproducible
- TODO (maybe): add the reference build to GitLab CI

schildbach commented:
At the moment, the following directory structure is produced:

```
$ tree output
output
├── bitcoinj-core-0.17-SNAPSHOT.jar
├── bitcoinj-core-0.17-SNAPSHOT-javadoc.jar
├── bitcoinj-core-0.17-SNAPSHOT-sources.jar
└── wallet-tool
    ├── bin
    │   ├── wallet-tool
    │   └── wallet-tool.bat
    └── lib
        ├── bcprov-jdk15to18-1.76.jar
        ├── bitcoinj-core-0.17-SNAPSHOT.jar
        ├── bitcoinj-wallettool.jar
        ├── checker-qual-3.33.0.jar
        ├── error_prone_annotations-2.18.0.jar
        ├── failureaccess-1.0.1.jar
        ├── guava-32.1.2-android.jar
        ├── j2objc-annotations-2.8.jar
        ├── jcip-annotations-1.0.jar
        ├── jsr305-3.0.2.jar
        ├── listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
        ├── picocli-4.7.4.jar
        ├── protobuf-javalite-3.22.3.jar
        ├── slf4j-api-2.0.7.jar
        └── slf4j-jdk14-2.0.7.jar
```

These are my checksums for the core artifacts, produced on x86_64 and Docker version 20.10.21:

```
$ sha256sum output/*.jar
cebd165603638cc3346152b6d85d81fe502c5f5bbd7d2d100fb973e9a26dfe31  output/bitcoinj-core-0.17-SNAPSHOT.jar
1f3959b0ac5284c70f81a3852b8d63e2aedd76ebbd1fb00340f2c0a6e2226b46  output/bitcoinj-core-0.17-SNAPSHOT-javadoc.jar
26102eeca8c329d06a32835dc151ddf400b170efb5373906580b9b443bbbb499  output/bitcoinj-core-0.17-SNAPSHOT-sources.jar
```

It would interesting to compare this to ARM architecture, or any other architecture that supports Docker/Podman and Debian.

schildbach commented:
The JavaDocs have ZIP files embedded, and those ZIP include timestamps. E.g.:

```
├── member-search-index.zip
│┄ Archive contents identical but files differ, possibly due to different compression levels. Falling back to binary comparison.
│ ├── zipinfo {}
│ │ @@ -1,3 +1,3 @@
│ │  Zip file size: 45652 bytes, number of entries: 1
│ │ --rw----     2.0 fat   411943 bl defN 23-Sep-17 11:58 member-search-index.json
│ │ +-rw----     2.0 fat   411943 bl defN 23-Sep-17 09:30 member-search-index.json
│ │  1 file, 411943 bytes uncompressed, 45490 bytes compressed:  89.0%
│ ├── member-search-index.zip
│ │ @@ -1,8 +1,8 @@
│ │ -00000000: 504b 0304 1400 0808 0800 525f 3157 0000  PK........R_1W..
│ │ +00000000: 504b 0304 1400 0808 0800 d54b 3157 0000  PK.........K1W..
```

msgilligan commented:
Running under **macOS** (M1/ARM) with **Docker Desktop** runs successfully, but produced different SHASUMs:

```
28f1d63edb276a9d64bd34ba348746485804084251e2c362edc171ae86a1a7b1  output/bitcoinj-core-0.17-SNAPSHOT.jar
623363588c360759dd11a6da72977c6e3c5c54613c9c47c117e0576aad35b8e2  output/bitcoinj-core-0.17-SNAPSHOT-javadoc.jar
f5b871c334e7ad29c18c2738ff1b1bb43e17fd7c1203f8410443d103d94dd151  output/bitcoinj-core-0.17-SNAPSHOT-sources.jar
```

schildbach commented:
I just fixed the output tree, it's now "wallet-tool/…" again. The output tree of the main artifacts just changed to a Maven Local Repo layout so it can include the POM file. For me, this step didn't change the checksums (apart from the JavaDoc which isn't reproducible atm).

schildbach commented:
> When I try to run this via **Podman** on **macOS** I get:
> 
> ```
> $ podman build --file build.Containerfile -o output .
> Error: '--output' option is not supported in remote mode
> ```

Yes, you need at least Podman 4 (I assume) for this option to be present. However, there is a workaround:

1. Run the command without the `-o`. It will create an image and print its ID.
2. `podman create <imageid> nop` (the command simply doesn't matter, we won't run the image)
3. `podman cp <containerid>:/ <outputdir>`
4. If you care, clean up the container/images

schildbach commented:
> but produced different SHASUMs

Can you attach the main JAR and the sources JAR to this issue? I'd like to run diffoscope.

schildbach commented:
> remote mode" means that Podman is using `podman machine` to "remotely" talk to a VM that is running Podman.

Why do you do this, is it necessary?

msgilligan commented:
Running **Docker** with `--platform linux/amd64` seems to have hung after completing most of the build:

```
$ docker build --platform linux/amd64 --file build.Containerfile -o output .
[+] Building 2540.8s (9/11)                                                                                                                        docker:desktop-linux
 => [internal] load build definition from build.Containerfile                                                                                                      0.0s
 => => transferring dockerfile: 1.05kB                                                                                                                             0.0s
 => [internal] load .dockerignore                                                                                                                                  0.0s
 => => transferring context: 222B                                                                                                                                  0.0s
 => [internal] load metadata for docker.io/library/debian:bullseye-slim                                                                                            2.0s
 => [auth] library/debian:pull token for registry-1.docker.io                                                                                                      0.0s
 => [build-stage 1/5] FROM docker.io/library/debian:bullseye-slim@sha256:3bc5e94a0e8329c102203c3f5f26fd67835f0c81633dd6949de0557867a87fac                          5.0s
 => => resolve docker.io/library/debian:bullseye-slim@sha256:3bc5e94a0e8329c102203c3f5f26fd67835f0c81633dd6949de0557867a87fac                                      0.0s
 => => sha256:3bc5e94a0e8329c102203c3f5f26fd67835f0c81633dd6949de0557867a87fac 1.85kB / 1.85kB                                                                     0.0s
 => => sha256:0cdee24a156b67861e0ec34a0784be80f3c6ec00f0d1708434395aa58d427d44 529B / 529B                                                                         0.0s
 => => sha256:316628d911722b9a861864b36cdaa1bf6cdacdfbd5900713eb919b01ab2ae409 1.46kB / 1.46kB                                                                     0.0s
 => => sha256:7d97e254a0461b0a30b3f443f1daa0d620a3cc6ff4e2714cc1cfd96ace5b7a7e 31.42MB / 31.42MB                                                                   3.8s
 => => extracting sha256:7d97e254a0461b0a30b3f443f1daa0d620a3cc6ff4e2714cc1cfd96ace5b7a7e                                                                          1.1s
 => [internal] load build context                                                                                                                                  0.1s
 => => transferring context: 195.53kB                                                                                                                              0.1s
 => [build-stage 2/5] RUN /usr/bin/apt-get update &&     /usr/bin/apt-get --yes install openjdk-11-jdk-headless gradle &&     /usr/sbin/adduser --disabled-logi  315.2s
 => [build-stage 3/5] WORKDIR /home/builder                                                                                                                        0.0s
 => [build-stage 4/5] COPY --chown=builder / project/                                                                                                              0.3s
 => [build-stage 5/5] RUN /usr/bin/gradle --project-dir project/     --no-build-cache --no-daemon --no-parallel     clean :bitcoinj-core:build :bitcoinj-walle  2218.3s 
 => => # :bitcoinj-wallettool:classes                                                                                                                                   
 => => # :bitcoinj-wallettool:jar                                                                                                                                       
 => => # :bitcoinj-wallettool:startScripts                                                                                                                              
 => => # :bitcoinj-wallettool:installDist                                                                                                                              
 => => # BUILD SUCCESSFUL in 8m 41s                                                                                                                                    
 => => # 23 actionable tasks: 19 executed, 4 up-to-date          
 ```

msgilligan commented:
> Can you attach the main JAR and the sources JAR to this issue? I'd like to run diffoscope.

Apparently not. GitHub says "We don't support that file type." I'll send to you via Matrix...

msgilligan commented:
> Yes, you need at least Podman 4

I'm using Podman 4.6.2 (this is a macOS build that only works in "remote" mode)

> However, there is a workaround:

I will try it.

> Why do you do this, is it necessary? The entire idea of Podman (as opposed to Docker) is that it can be simply run as a tool and as a simple user, without any infrastructure (daemon process, etc).

Linux containers require a Linux kernel. So both Docker and Podman require the use of a Linux VM. I'm not sure why Docker is able to make `--output` work in remote mode and Podman isn't. Maybe they just haven't implemented it yet.

> Anyhow, you could try "buildah" instead, which is limited to just building images but that's exactly what we need here.

I haven't used "buildah" before, but will look into it.

msgilligan commented:
`buildah` doesn't seem to be available for macOS or Windows.

msgilligan reviewed (changes requested):
This is a preliminary review of what I believe is a preliminary version -- since we haven't resolved the unexplained meta-data differences we are seeing.

msgilligan commented on build.Containerfile:
```diff
@@ -0,0 +1,39 @@
+# prepare debian environment
+ENV DEBIAN_FRONTEND noninteractive
+RUN /usr/bin/apt-get update && \
+    /usr/bin/apt-get --yes install openjdk-11-jdk-headless gradle && \
```
I'm also hoping (though it is not critical) that we can use bookworm and JDK 17 here.

schildbach replied:
Atm I wanted to resemble the status quo. The switch from JDK 11 to 17 is orthogonal.

schildbach commented on build.Containerfile:
```diff
@@ -0,0 +1,39 @@
+RUN /usr/bin/gradle --project-dir project/ \
+    --no-build-cache --no-daemon --no-parallel \
+    -Dmaven.repo.local=repo \
+    clean :bitcoinj-core:publishToMavenLocal :bitcoinj-wallettool:installDist
```
1. it's for getting the POM with a nice name
2. the structure is close resembles what we upload to Maven Central, so I thought it's a good fit

schildbach replied:
Does this answer your question?

msgilligan replied:
It answers my question, but I'd prefer if we just used the direct outputs of the primary build targets as the reproducible outputs.

msgilligan replied:
To avoid making an unnecessary copy of the files and to focus our verification efforts on the normal build outputs.

And there's a philosophical (and possibly practical) objection. Publishing to maven local is technically a side effect and mutates the developers system.  (This is in the context of "functional-style" build tools like Nix and Guix.)  I know you have it configured to write to the local directory, but using `publishToMavenLocal` just "feels wrong" and makes me nervous. This may just be superstition on my part, though.

schildbach replied:
By the way, the "publish" tasks don't run our tests. This can be seen as bad but also as good. In fact, reproducable builds is *not* about correctness of code, it's just about reproducibility. Since we have tons of tests for correctness, I thought not running the tests in our reproducable build is nice to have.

However, I'm not adamant on omitting the tests. We could add the `:bitcoinj-core:build` or `:bitcoinj-core:test` tasks to the gradle line. However, I'm pretty sure I will not be able to verify on my Raspi Zero 2W any more if we do…

msgilligan replied:
> I think we have some mismatch of understanding of what reproducible builds (or maybe verifyable builds?) is about. 

I don't think so. I think we just differ over what should be the final output.

I would like a regular. local build to produce output that is reproducible and to be able to easily verify the shasums. We pretty much have that now, so maybe this doesn't matter.

msgilligan replied:
> By the way, the "publish" tasks don't run our tests. 

I noticed this. I suppose I would feel better if the tests were run as part of the process of producing an official image, but I suppose the tests could be run separately and the shasums of the two images compared so that we know the official image passed the tests.

Perhaps there could be a config flag to turn on/off testing in the reference build?

schildbach commented on core/build.gradle:
```diff
@@ -52,9 +52,10 @@ protobuf {
+tasks.withType(Zip) {
     preserveFileTimestamps = false
     reproducibleFileOrder = true
+    entryCompression = ZipEntryCompression.STORED
```
Did you manage to reproduce this uncompressed version?

msgilligan replied:
Yes, but it didn't reproduce the SHA256 sums you sent me. I got:
```
$ shasum -a 256 *.jar
6615bc457188c42afeb3a1eed965be9a6825a53feab4b7956372ea6317580a04  bitcoinj-core-0.17-SNAPSHOT-javadoc.jar
d933f7df8d98c43dd63746bc5c3d58d0e212efb535c7bb48962c7eab7c1e2169  bitcoinj-core-0.17-SNAPSHOT-sources.jar
21c9c1aaf695b7411262836e1684408d871cefd0719246bcaa0b40a043afa162  bitcoinj-core-0.17-SNAPSHOT.jar
```

msgilligan commented:
> Running under **macOS** (M1/ARM) with **Docker Desktop** runs successfully, but produced different SHASUMs:

This was partly explained by me forgetting that my working-dirs were not clean. And partly explained by the differences between meta-data we are trying to explain.

schildbach commented:
I just added normalization of the project file perms (f93bac950ed444eb51a635648235c52df7bf8f78) and now we're bit-identical!

schildbach commented:
Then, I've enabled compression again (051e249f49e252f31f27245e19ae9859b1079c65).

The checksums for me are now:

```
6d2e93c56c4ab59885d4b2d623592c86d2cc347c4d2aba029d83af37386988a5  build/output/org/bitcoinj/bitcoinj-core/0.17-SNAPSHOT/bitcoinj-core-0.17-SNAPSHOT.jar
6b59515c2498c550ac209cb89da4917c2f8c5a401ddc40d322d096c5108acc45  build/output/org/bitcoinj/bitcoinj-core/0.17-SNAPSHOT/bitcoinj-core-0.17-SNAPSHOT-javadoc.jar
d835c79d69bc811914bac3b9befb09b039e521e087cc367e18579e5cffcaebde  build/output/org/bitcoinj/bitcoinj-core/0.17-SNAPSHOT/bitcoinj-core-0.17-SNAPSHOT-sources.jar
```

schildbach commented:
After some experimentation, I've flattened the output directory structure a bit to:

```
output
├── core
│   ├── bitcoinj-core-0.17-SNAPSHOT.jar
│   ├── bitcoinj-core-0.17-SNAPSHOT-javadoc.jar
│   ├── bitcoinj-core-0.17-SNAPSHOT.pom
│   └── bitcoinj-core-0.17-SNAPSHOT-sources.jar
└── wallettool
    ├── bin
    │   ├── wallet-tool
    │   └── wallet-tool.bat
    └── lib
        ├── bcprov-jdk15to18-1.76.jar
        ├── bitcoinj-core-0.17-SNAPSHOT.jar
        ├── bitcoinj-wallettool.jar
        ├── checker-qual-3.33.0.jar
        ├── error_prone_annotations-2.18.0.jar
        ├── failureaccess-1.0.1.jar
        ├── guava-32.1.2-android.jar
        ├── j2objc-annotations-2.8.jar
        ├── jcip-annotations-1.0.jar
        ├── jsr305-3.0.2.jar
        ├── listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
        ├── picocli-4.7.4.jar
        ├── protobuf-javalite-3.22.3.jar
        ├── slf4j-api-2.0.7.jar
        └── slf4j-jdk14-2.0.7.jar
```

msgilligan commented on build.Containerfile:
```diff
@@ -0,0 +1,43 @@
+# prepare debian environment
+ENV DEBIAN_FRONTEND noninteractive
+RUN /usr/bin/apt-get update && \
+    /usr/bin/apt-get --yes install openjdk-11-jdk-headless gradle && \
```
This is potentially non-reproducible because the versions are not locked. In practice, however, I doubt it will be an issue.

schildbach replied:
That's true. It's hard to strike the right balance between security patches and reproducibility. However I also don't expect much trouble here.

Sadly there is no Debian variant on Docker Hub that doesn't have `-updates` enabled.

schildbach commented:
I think we're almost good to merge this.

> * FIXME the JavaDoc doesn't seem to be reproducible

Since this is likely rooted in the JDK (Doclet) and we're going to upgrade the JDK anyway, I'll defer this.

>     * TODO (maybe): add the reference build to GitLab CI

I'll defer this to a followup PR as well. I've got the necessary code on another repo already.

schildbach commented:
> Then, I've enabled compression again ([051e249](https://github.com/bitcoinj/bitcoinj/commit/051e249f49e252f31f27245e19ae9859b1079c65)).
> 
> The checksums for me are now:
> 
> ```
> 6d2e93c56c4ab59885d4b2d623592c86d2cc347c4d2aba029d83af37386988a5  build/output/org/bitcoinj/bitcoinj-core/0.17-SNAPSHOT/bitcoinj-core-0.17-SNAPSHOT.jar
> 6b59515c2498c550ac209cb89da4917c2f8c5a401ddc40d322d096c5108acc45  build/output/org/bitcoinj/bitcoinj-core/0.17-SNAPSHOT/bitcoinj-core-0.17-SNAPSHOT-javadoc.jar
> d835c79d69bc811914bac3b9befb09b039e521e087cc367e18579e5cffcaebde  build/output/org/bitcoinj/bitcoinj-core/0.17-SNAPSHOT/bitcoinj-core-0.17-SNAPSHOT-sources.jar
> ```

@msgilligan , can you check on your system if your shasums match? (Except for the Javadoc, which are known to include a timestamp.)

schildbach commented:
I replaced my `chmod` workaround with the `fileMode` / `dirMode` properties from #3274 – it seems the cleaner way to do this (as long as there are no executables within archives).

schildbach commented:
Just to let you know, the recent commits didn't change the `sha256sum` of the JARs for me, so my last post is still valid.

msgilligan commented:
> @msgilligan , can you check on your system if your shasums match? (Except for the Javadoc, which are known to include a timestamp.)

Yes, the classfile JAR and sources JAR match.  (The JavaDoc JAR as expected does not match.)

schildbach commented:
Rebased & squashed, and I've added an extensive commit comment:

```
    build.Containerfile: reproducible reference build via containerization
    
    Currently, Docker and Podman are supported.
    
    The goals of the reference build are:
    
    - Make our release process more transparent
    - Improve reproducibility by shielding our build from the hosts
      environment
    - Reduce the trust necessary for running our build by shielding
      the hosts environment from possible side effects of our build
    
    Note both the signing and the uploading of release artifacts are out
    of scope for the reference build.
```

msgilligan commented:
When I try to run this on ARM64 macOS in a x86 VM, the build hangs as follows:
```
$ docker build --platform linux/amd64 --file build.Containerfile --output output .
[+] Building 256.6s (9/12)                                                                                          docker:desktop-linux
 => [internal] load build definition from build.Containerfile                                                                       0.0s
 => => transferring dockerfile: 1.20kB                                                                                              0.0s
 => [internal] load .dockerignore                                                                                                   0.0s
 => => transferring context: 222B                                                                                                   0.0s
 => [internal] load metadata for docker.io/library/debian:bullseye-slim                                                             1.6s
 => [auth] library/debian:pull token for registry-1.docker.io                                                                       0.0s
 => [build-stage 1/5] FROM docker.io/library/debian:bullseye-slim@sha256:3bc5e94a0e8329c102203c3f5f26fd67835f0c81633dd6949de055786  0.0s
 => [internal] load build context                                                                                                   0.0s
 => => transferring context: 51.33kB                                                                                                0.0s
 => CACHED [build-stage 2/5] RUN /usr/bin/apt-get update &&     /usr/bin/apt-get --yes install openjdk-11-jdk-headless gradle &&    0.0s
 => CACHED [build-stage 3/5] WORKDIR /home/builder                                                                                  0.0s
 => [build-stage 4/5] COPY --chown=builder / project/                                                                               0.1s
 => [build-stage 5/5] RUN /usr/bin/gradle --project-dir project/     --no-build-cache --no-daemon --no-parallel     -Dmaven.repo  254.8s
 => => # :bitcoinj-wallettool:classes                                                                                                   
 => => # :bitcoinj-wallettool:jar                                                                                                       
 => => # :bitcoinj-wallettool:startScripts                                                                                              
 => => # :bitcoinj-wallettool:installDist                                                                                               
 => => # BUILD SUCCESSFUL in 2m 44s                                                                                                     
 => => # 19 actionable tasks: 15 executed, 4 up-to-date    
 ```

schildbach commented:
> When I try to run this on ARM64 macOS in a x86 VM, the build hangs as follows:
> 
> ```
> $ docker build --platform linux/amd64 --file build.Containerfile --output output .
> [+] Building 256.6s (9/12)                                                                                          docker:desktop-linux
>  => [internal] load build definition from build.Containerfile                                                                       0.0s
>  => => transferring dockerfile: 1.20kB                                                                                              0.0s
>  => [internal] load .dockerignore                                                                                                   0.0s
>  => => transferring context: 222B                                                                                                   0.0s
>  => [internal] load metadata for docker.io/library/debian:bullseye-slim                                                             1.6s
>  => [auth] library/debian:pull token for registry-1.docker.io                                                                       0.0s
>  => [build-stage 1/5] FROM docker.io/library/debian:bullseye-slim@sha256:3bc5e94a0e8329c102203c3f5f26fd67835f0c81633dd6949de055786  0.0s
>  => [internal] load build context                                                                                                   0.0s
>  => => transferring context: 51.33kB                                                                                                0.0s
>  => CACHED [build-stage 2/5] RUN /usr/bin/apt-get update &&     /usr/bin/apt-get --yes install openjdk-11-jdk-headless gradle &&    0.0s
>  => CACHED [build-stage 3/5] WORKDIR /home/builder                                                                                  0.0s
>  => [build-stage 4/5] COPY --chown=builder / project/                                                                               0.1s
>  => [build-stage 5/5] RUN /usr/bin/gradle --project-dir project/     --no-build-cache --no-daemon --no-parallel     -Dmaven.repo  254.8s
>  => => # :bitcoinj-wallettool:classes                                                                                                   
>  => => # :bitcoinj-wallettool:jar                                                                                                       
>  => => # :bitcoinj-wallettool:startScripts                                                                                              
>  => => # :bitcoinj-wallettool:installDist                                                                                               
>  => => # BUILD SUCCESSFUL in 2m 44s                                                                                                     
>  => => # 19 actionable tasks: 15 executed, 4 up-to-date    
> ```
> 
> Since running in the default ARM64 mode seems to work properly, this isn't a huge issue. But its possible this could happen in other configurations.

Try adding `--progress plain` to the Docker command line. Does it say `DONE` on the `RUN /usr/bin/gradle` action?

schildbach commented:
I just confirmed that running this reference build on arm64 (Raspi Pi Zero 2W running Docker version 24.0.5, build 24.0.5-0ubuntu1~22.04.1) yields the correct checksums!

However, the build took about 30 minutes and therefore exceeds some weird timeout for waiting on the final context to export. Running the build twice works around this because by re-using the caches the second invocation only takes 5 seconds.

```
# docker build --file build.Containerfile -o build/output .
[+] Building 1770.5s (12/12) FINISHED                                                                                                                         docker:default
 => [internal] load build definition from build.Containerfile                                                                                                           0.2s
 => => transferring dockerfile: 1.20kB                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                       0.2s
 => => transferring context: 222B                                                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/debian:bullseye-slim                                                                                                 2.9s
 => [build-stage 1/5] FROM docker.io/library/debian:bullseye-slim@sha256:c618be84fc82aa8ba203abbb07218410b0f5b3c7cb6b4e7248fda7785d4f9946                              23.8s
 => => resolve docker.io/library/debian:bullseye-slim@sha256:c618be84fc82aa8ba203abbb07218410b0f5b3c7cb6b4e7248fda7785d4f9946                                           0.1s
 => => sha256:c618be84fc82aa8ba203abbb07218410b0f5b3c7cb6b4e7248fda7785d4f9946 1.85kB / 1.85kB                                                                          0.0s
 => => sha256:fe434fe10d82fb78bb8a9819f19a854c2d17ef87f2f7794ea0155bdb5ad89f08 529B / 529B                                                                              0.0s
 => => sha256:3fa8a56f7bc99978173e7bd6de8e4d6112bcc40575c57efb01c7d690656ac741 1.48kB / 1.48kB                                                                          0.0s
 => => sha256:fc521c5b98350f6fd8c72ace1e48558bb7b53cb3db201a2a3b42095401cd02f1 30.06MB / 30.06MB                                                                        8.9s
 => => extracting sha256:fc521c5b98350f6fd8c72ace1e48558bb7b53cb3db201a2a3b42095401cd02f1                                                                              13.1s
 => [internal] load build context                                                                                                                                       2.6s
 => => transferring context: 9.02MB                                                                                                                                     2.4s
 => [build-stage 2/5] RUN /usr/bin/apt-get update &&     /usr/bin/apt-get --yes install openjdk-11-jdk-headless gradle &&     /usr/sbin/adduser --disabled-login --g  230.4s
 => [build-stage 3/5] WORKDIR /home/builder                                                                                                                             0.9s
 => [build-stage 4/5] COPY --chown=builder / project/                                                                                                                   3.3s
 => [build-stage 5/5] RUN /usr/bin/gradle --project-dir project/     --no-build-cache --no-daemon --no-parallel     -Dmaven.repo.local=repo     clean :bitcoinj-cor  1487.2s
 => [export-stage 1/2] COPY --from=build-stage     /home/builder/repo/org/bitcoinj/*/*/*.jar     /home/builder/repo/org/bitcoinj/*/*/*.pom     /core/                   0.6s 
 => [export-stage 2/2] COPY --from=build-stage     /home/builder/project/wallettool/build/install/wallet-tool/     /wallettool/                                         1.0s 
 => ERROR exporting to client directory                                                                                                                                 5.0s 
------                                                                                                                                                                       
 > exporting to client directory:                                                                                                                                            
------                                                                                                                                                                       
ERROR: failed to solve: DeadlineExceeded: no active session for vfcmaf0riatxfycpyc6zns2pz: context deadline exceeded

# ls -la build/output
total 8
drwxr-xr-x 2 root root 4096 Sep 20 09:17 .
drwxr-xr-x 3 root root 4096 Sep 20 09:17 ..

# docker build --file build.Containerfile -o build/output .
[+] Building 5.2s (12/12) FINISHED                                                                                                                            docker:default
 => [internal] load .dockerignore                                                                                                                                       0.1s
 => => transferring context: 222B                                                                                                                                       0.1s
 => [internal] load build definition from build.Containerfile                                                                                                           0.2s
 => => transferring dockerfile: 1.20kB                                                                                                                                  0.1s
 => [internal] load metadata for docker.io/library/debian:bullseye-slim                                                                                                 1.9s
 => [build-stage 1/5] FROM docker.io/library/debian:bullseye-slim@sha256:c618be84fc82aa8ba203abbb07218410b0f5b3c7cb6b4e7248fda7785d4f9946                               0.0s
 => [internal] load build context                                                                                                                                       0.7s
 => => transferring context: 51.19kB                                                                                                                                    0.6s
 => CACHED [build-stage 2/5] RUN /usr/bin/apt-get update &&     /usr/bin/apt-get --yes install openjdk-11-jdk-headless gradle &&     /usr/sbin/adduser --disabled-logi  0.0s
 => CACHED [build-stage 3/5] WORKDIR /home/builder                                                                                                                      0.0s
 => CACHED [build-stage 4/5] COPY --chown=builder / project/                                                                                                            0.0s
 => CACHED [build-stage 5/5] RUN /usr/bin/gradle --project-dir project/     --no-build-cache --no-daemon --no-parallel     -Dmaven.repo.local=repo     clean :bitcoinj  0.0s
 => CACHED [export-stage 1/2] COPY --from=build-stage     /home/builder/repo/org/bitcoinj/*/*/*.jar     /home/builder/repo/org/bitcoinj/*/*/*.pom     /core/            0.0s
 => CACHED [export-stage 2/2] COPY --from=build-stage     /home/builder/project/wallettool/build/install/wallet-tool/     /wallettool/                                  0.0s
 => exporting to client directory                                                                                                                                       2.0s
 => => copying files 16.56MB                   

# ls -la build/output
total 16
drwxr-xr-x 4 root root 4096 Sep 20 10:12 .
drwxr-xr-x 3 root root 4096 Sep 20 09:17 ..
drwxr-xr-x 2 root root 4096 Sep 20 10:09 core
drwxr-xr-x 4 root root 4096 Sep 20 10:09 wallettool

# sha256sum build/output/core/*
58aed2d39d47c77e4c59489022f79e613482d2baf6e2ac4aa7fc46b456907130  build/output/core/bitcoinj-core-0.17-SNAPSHOT-javadoc.jar
d835c79d69bc811914bac3b9befb09b039e521e087cc367e18579e5cffcaebde  build/output/core/bitcoinj-core-0.17-SNAPSHOT-sources.jar
6d2e93c56c4ab59885d4b2d623592c86d2cc347c4d2aba029d83af37386988a5  build/output/core/bitcoinj-core-0.17-SNAPSHOT.jar
22977f9e0be235a6f7a6ce03d10293955f718e8e331e984a43e4aae5ca4657f0  build/output/core/bitcoinj-core-0.17-SNAPSHOT.pom
```

schildbach commented:
I clarified the requirements in the readme and the comments section of the containerfile a bit, introducing buildah. No functional changes.

msgilligan commented on README.adoc:
```diff
@@ -74,6 +74,12 @@ To dump the state of the wallet in `~/bitcoinj/bitcoinj-test.wallet` with the te
+
+Our reference build (which is also used for our releases) is running within a container to provide good reproducibility.
+Buildah 1.26+, Podman 4.1+ and Docker (with BuildKit) are supported. We tested many combinations of these host OSes:
+Debian, Ubuntu, MacOS, Windows (WSL); and these architectures: amd64, arm64. For instructions see `build.Containerfile`.
```
on macOS you must use Docker because Podman doesn't work and Buildah is Linux-only.

schildbach replied:
I see, thanks. I didn't list all working combinations because I thought it's too much of information. Maybe we should recommend one working solution per OS, but we can defer this until people start using the reference build. I'll mark this as resolved.

schildbach commented:
I just added a `setup-stage`. It didn't even increase build time – I expected much more of a hit. Example usage:

```
$ docker build --file build.Containerfile --target setup-stage --tag mydebian .
[+] Building 2.0s (6/6) FINISHED                                                                                                                                                                                                                                                                                                            docker:default
 => [internal] load .dockerignore                                                                                                                                                                                                                                                                                                                     0.1s
 => => transferring context: 222B                                                                                                                                                                                                                                                                                                                     0.0s
 => [internal] load build definition from build.Containerfile                                                                                                                                                                                                                                                                                         0.1s
 => => transferring dockerfile: 1.33kB                                                                                                                                                                                                                                                                                                                0.0s
 => [internal] load metadata for docker.io/library/debian:bullseye-slim                                                                                                                                                                                                                                                                               0.0s
 => [setup-stage 1/2] FROM docker.io/library/debian:bullseye-slim                                                                                                                                                                                                                                                                                     0.0s
 => CACHED [setup-stage 2/2] RUN /usr/bin/apt-get update &&     /usr/bin/apt-get --yes install openjdk-11-jdk-headless gradle &&     /usr/sbin/adduser --disabled-login --gecos "" builder                                                                                                                                                            0.0s
 => exporting to image                                                                                                                                                                                                                                                                                                                                1.6s
 => => exporting layers                                                                                                                                                                                                                                                                                                                               1.6s
 => => writing image sha256:f5115d61dbbb71c4a93cbb3f0317ebb62a67ceff42095a9e0b863c0de17d3a32                                                                                                                                                                                                                                                          0.0s
 => => naming to docker.io/library/mydebian                                                                                                                                                                                                                                                                                                           0.0s

$ docker run mydebian gradle --version
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/usr/share/java/groovy-all.jar) to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

------------------------------------------------------------
Gradle 4.4.1
------------------------------------------------------------

Build time:   2012-12-21 00:00:00 UTC
Revision:     none

Groovy:       2.4.21
Ant:          Apache Ant(TM) version 1.10.9 compiled on December 25 1969
JVM:          11.0.20 (Debian 11.0.20+8-post-Debian-1deb11u1)
OS:           Linux 6.2.0-33-generic amd64
```

schildbach commented:
And finally, I've added a build-arg to add an addtional gradle tasks. Use like this:

```
docker build --file build.Containerfile --build-arg ADDITIONAL_GRADLE_TASK=test --output <outputdir> .
```

schildbach commented:
By the way, I plan to experiment with bootstrapping a Debian with a couple of commands (`FROM scratch`), rather than using the image from docker hub. This would put us in control over what kind of patches we want (ideally just security). It would also change the trust anchor, for better or worse…

It out of scope for this PR however.

msgilligan reviewed (approved):
Approved!
